### ISSUE TITLE
c-ares resolver blocks channel indefinitely if partial or no response received from DNS

### ISSUE THREAD

**Issue by technicianted** [grpc/grpc#17436]
### What version of gRPC and what language are you using?
 1.16
 
### What operating system (Linux, Windows, …) and version?
Linux - Ubuntu 16.04
 
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10)
 
 
### What did you do?
```
export GRPC_DNS_RESOLVER=ares
```
change `greeter_client.cc`:
```c++
  GreeterClient greeter(grpc::CreateChannel(
      "dns://172.217.3.174:53/example.com", grpc::InsecureChannelCredentials()));
```
 
where `172.217.3.174` is partially or fully unreachable IP address.

### What did you expect to see?
 DNS resolution attempt would time out and call would fail.
 
### What did you see instead?
Resolution was tried once and never tried again, and the resolution never timed out blocking call indefinitely.

You can see from the trace output that ares never tried again:
```
D1206 15:32:24.060189349   90518 dns_resolver_ares.cc:175]   AresDnsResolver::NextLocked() is called.
D1206 15:32:24.060193149   90518 dns_resolver_ares.cc:415]   Start resolving.
D1206 15:32:24.060335349   90518 grpc_ares_ev_driver.cc:128] grpc_ares_ev_driver_create_locked
I1206 15:32:24.060505148   90518 ev_posix.cc:245]            (fd-trace) fd_create(5, c-ares fd: 5, 0)
D1206 15:32:24.060519448   90518 grpc_ares_ev_driver.cc:255] new fd: c-ares fd: 5
D1206 15:32:24.060524848   90518 grpc_ares_ev_driver.cc:85]  Ref ev_driver 11563248
D1206 15:32:24.060528948   90518 grpc_ares_ev_driver.cc:272] notify read on: c-ares fd: 5
E1206 15:32:24.060534548   90518 lockfree_event.cc:98]       LockfreeEvent::NotifyOn: 0xb0a7b8 curr=(nil) closure=0xb0a668
I1206 15:32:24.060540048   90518 client_channel.cc:2857]     chand=0xb07d70 calld=0xb08fc0: deferring pick pending resolver result
```

Then no forward progress.

Same behavior is observed if no response received for only one of the queries (`A`, `AAAA`, `SRV`, `TXT`).

**Comment by apolcyn** [comment]
this should be fixed by https://github.com/grpc/grpc/pull/17382, which will also be released in a 1.17.1 soon

**Comment by technicianted** [comment]
Thanks. I'm curious on why you chose to use an independent timer rather than use ares timeouts?

**Comment by apolcyn** [comment]
The timeout in this case happens while grpc library poller is waiting for an event on a c-ares fd, c-ares  library itself doesn't really have a chance to take any action during this time.

**Comment by technicianted** [comment]
I meant a grpc timer on c-ares `ares_timeout()` interval to `ares_process()`.

**Comment by apolcyn** [comment]
hmm... the existing fix was really chosen mostly for simplicity.  But as a hypothetical scenario, I suppose what you are describing may be a good way of handling e.g. cases in which we want c-ares to iterate over a server list, to go past faulty DNS servers at the beginning of the list. Is this why you are suggesting this? Just wondering here.... are you seeing or a real (or hypothetical) issue which would be solved by your suggestion but not by the existing fix?

**Comment by technicianted** [comment]
The existing fix will not apply conventional DNS retries for partially faulty or overloaded servers.

In my use-case I'm being hit by something similar to [this](https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts) (especially common in kubernetes services), where some dns udp packets get dropped due to race condition in NAT, causing some of c-ares queries not to make it to the server. It is especially bad for c-are + grpc since it sends (at least) 4 queries one shot.

Your solution will break the wait and error out. But what you would expect from a dns client is to eventually succeed.

One other minor point is by using c-ares timeouts you get it to behave like it is documented. For example, you can control timeouts and retries from env variables.

**Comment by apolcyn** [comment]
https://github.com/grpc/grpc/pull/17688 in progress FYI - I think this suggestion works well, but making one simplifying tweak to skip the `ares_timeout` calls.

**Comment by apolcyn** [comment]
The issue initially posted about was fixed in #17382, the issue described in https://github.com/grpc/grpc/issues/17436#issuecomment-445423269 was addressed by https://github.com/grpc/grpc/pull/17688
